File: 66/plus_one.py
"""
Given a non-negative number represented as an array of digits, plus one to the number.

The digits are stored such that the most significant digit is at the head of the list."""

class Solution(object):

    def plusOne(self, digits):

        """

        :type digits: List[int]

        :rtype: List[int]

        """

        

        # traverse in reverse

        plus_one = True

        for i in range(len(digits))[::-1]:

            if plus_one:

                digits[i] = (digits[i]+1)%10

            if digits[i] != 0:

                plus_one = False

                

        # overflow

        if plus_one:

            ret = [1]

            for d in digits:

                ret.append(d)

            return ret

                

        # return new list

        return [digit for digit in digits]

File: 66/plus_one.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        return plusN(digits, 1);
    }
    
    vector<int> plusN(vector<int>& digits, int n) {
        vector<int> ret;
        
        int carry = n;
        for (int i=digits.size()-1; i>=0; i--) {
            ret.insert(ret.begin(), (digits[i] + carry)%10);
            if (carry!=0)
                carry = (digits[i] + 1)/10;
        }
        
        if (carry!=0) {
            ret.insert(ret.begin(), carry);
        }
        
        return ret;
    }
};